### DISCUSSION TITLE
Some questions about parking_env

### DISCUSSION THREAD

**Discussion by xunpla123** [eleurent/highway-env#173] [General]
I'm implementing the parking_env and have done some other tests. I added a mark to the front (or rear) of the car and found that the car always reaches the goal with the same side (front or rear of the car). I think that's the front haha.

And i have two questions now :
1) If the car actually reaches the goal with the same side, how do you do that ?
2) I try to find  some descriptions about the front and rear of the car (not front car and rear car) , can you tell me which part of the code are they in ?

Looking forward to your reply, thanks.

**Comment by eleurent** [comment]
It is true that the front and read of the car cannot really be distinguished from the graphics right now...

1. in `parking_env`, at the moment the car is tasked to always reach the goal from the front side because the [goal heading is set to the lane heading](https://github.com/eleurent/highway-env/blob/master/highway_env/envs/parking_env.py#L88) (outwards). You can make it to switch randomly to the rear side, for instance, by setting: ```self.goal = Landmark(self.road, lane.position(lane.length/2, 0), heading=lane.heading + self.np_random.randint(2)*np.pi)```
I may add this option and make it configurable.
2. The front/rear of the car are only distinguished in the code by the [vehicle heading](https://github.com/eleurent/highway-env/blob/master/highway_env/vehicle/kinematics.py#L44). You can edit the [vehicle display ](https://github.com/eleurent/highway-env/blob/master/highway_env/vehicle/graphics.py#L28)method as you wish to show this information (e.g. draw a windshield).

Does that help?

**Comment by xunpla123** [comment]
Thanks for your reply, this answer really helps me a lot. Then, i have a new question now haha !

I hope my video shows the car ’s whole trajectory, and i found the functions, [display_trajectory](https://github.com/eleurent/highway-env/blob/d8de1498b3e9ac0f67726bc3dbafea4f07028b4f/highway_env/vehicle/graphics.py#L107) and [display_history](https://github.com/eleurent/highway-env/blob/d8de1498b3e9ac0f67726bc3dbafea4f07028b4f/highway_env/vehicle/graphics.py#L119).
But i don't know how to implement the functions in `parking_env`.

**Comment by eleurent** [comment]
[display_history](https://github.com/eleurent/highway-env/blob/d8de1498b3e9ac0f67726bc3dbafea4f07028b4f/highway_env/vehicle/graphics.py#L119) is used to show the *past* trajectory of a vehicle. It can be enabled by using the following environment configuration:
```
env.configure({"show_trajectories": True})
```

[display_trajectory](https://github.com/eleurent/highway-env/blob/d8de1498b3e9ac0f67726bc3dbafea4f07028b4f/highway_env/vehicle/graphics.py#L107), however, shows the predicted future trajectory *given a sequence of actions to perform*. In order to enable it, you must call
```
env.viewer.set_agent_action_sequence(actions)
env.step(actions[0])
env.render()
```
where `actions` is the list of actions that you plan to execute.

**Comment by xunpla123** [comment]
Thanks for your answers. It's really a great environment !
